### PR TITLE
fix: a cube constraint shall constrain its observations

### PR DESCRIPTION
**Pull request by Rdataflow**
@ktk a cube constraint shall constrain its observations 

according to the cube specs in https://cube.link/#observationConstraint

cc @kronmar @tpluscode @BenjaminHofstetter @claudiodigallo

### CODE REVIEW

File: validation/standalone-constraint-constraint.ttl [hunk 1]
```diff
@@ -17,6 +17,12 @@
 <ObservationConstraintShape>
     a sh:NodeShape ;
     sh:targetClass cube:Constraint ;
+    sh:property [
+        # a cube constraint shall constrain its observations
+        sh:path sh:targetClass ;
+        sh:hasValue cube:Observation;
+        sh:message "cube:Constraint needs a sh:targetClass cube:Observation"
+    ] ;
```
Comment (tpluscode) [inline validation/standalone-constraint-constraint.ttl:25]: Every cube's observations have that class. When we start adding `sh:targetClass cube:Constraint` to cube shapes these shapes will not only constrain their own observations but ALL observations across all cubes.

See [here](https://s.zazuko.com/2tWqa9v) how the dimensions are constrained across cubes

Comment (Rdataflow) [reply validation/standalone-constraint-constraint.ttl:25]: @tpluscode 💯 true
@kronmar can you jump in an describe your observed issue more precisely ... THX

**Comment by tpluscode** [comment]
If I understand the requirements, I see two options for what can be done here, both on per-cube basis

First is the use of SPARQL Targets. A SPARQL Target is a SPARQL query attached to the shape which lets engines find the observations. 

```turtle
<Cube/shape>
  a cube:Constraint ;
  sh:target [
    a sh:SPARQLTarget ;
    sh:prefixes [
		sh:declare [
			sh:namespace <https://cube.link/> ;
			sh:prefix "cube" ;
		] ;
	] ;
	sh:select "SELECT ?this WHERE { <Cube> cube:observationSet/cube:observation ?this . }" ;	
  ] ;
.
```

Stardog does not appear to support them but can be used to validate using Jena for example. 

An alternative solution would be a specialised type for cube's observations. A subclass of `cube:Observation` would need to be added for a particular cube.

```turtle
<Cube/shape>
  sh:target <Cube/Observation> ;
.

# not necessary unless we use reasoning but might keep to make it explicit
<Cube/Observation> rdfs:subClassOf cube:Observation .

<Cube/observation/1>
  a cube:Observation, <Cube/Observation>
.
```

**Comment by kronmar** [comment]
Or a third one, but that would be very verbose [see heere](https://s.zazuko.com/NnS5P6):
```
<Cube/shape>
  a cube:Constraint ;
  sh:targetNode <Cube/observation/1>, <Cube/observation/2>, ... , <Cube/observation/N> 
.
```` 
But then again, the number of triples needed to be added is the same as in your subClass solution.

**Comment by Rdataflow** [comment]
... and a fourth solution is to introduce the *missing link*:
i.e.

```ttl
<temperature-sensor/cube/missing-link> a sh:NodeShape ;
  sh:targetNode <temperature-sensor/cube> ;
  sh:property [
    sh:path cube:observationSet ;
    sh:node [
      sh:property [
        sh:path cube:observation ;
        sh:node <temperature-sensor/cube/constraint> ;
      ]
    ]
  ]
  .
```

shacl validation fails then as expected https://s.zazuko.com/VLDoy4

*nb: the validation report in shacl-playground doesn't show the nested errors... a thing that is already fixed in the CLI tool https://github.com/zazuko/cube-link/commit/4d251a4c99ef634401ef3d65c89a3417d54e2e82*

**Comment by giacomociti** [comment]
@Rdataflow apparently your fourth solution is already in place since long ago, except it is not working due to a naming issue: https://github.com/zazuko/cube-link/commit/dfbfe5a390a38c9f231a98d15b09480eff4babf2

**Comment by Rdataflow** [comment]
@giacomociti that's great THX so we look forward to have #111 merged 👍

cc @kronmar validation of values fix is on the road within validation, so there is no need to alter any data 🎉

**Comment by kronmar** [comment]
So for my understanding, @giacomociti , these missing triples are added during the validation step?

**Comment by giacomociti** [comment]
> So for my understanding, @giacomociti , these missing triples are added during the validation step?

exactly

**Comment by kronmar** [comment]
Ah okay.

Could you document that on [cube.link](https://cube.link/) to avoid confusion about that?

**Comment by Rdataflow** [comment]
closed by #111

**Pull request closed without merging**